### DISCUSSION TITLE
MDX not compiling when using React Component in mdx file

### DISCUSSION THREAD

**Discussion by Yusufzai** [vercel/next.js#58882] [Help]
### Summary

I am using NextJS and MDX with next-mdx-remote to get my data and it successfully work but the only problem is whenever I try to import react components I get the following error

> Error: Expected component CodeBlock to be defined: you likely forgot to import, pass, or provide it. It’s referenced in your code at 317:1-317:24

In short :

This Works -

```
---
title: 'First Blog Post'
author: 'Me'
publishDate: 'Nov 4th, 2023'
---

# This is our first post

Here is a regular paragraph

here is a list:

- Item one
- Item two
- Item three

## This is an h2 element

```

This Does Not -

```
---
title: 'First Blog Post'
author: 'Me'
publishDate: 'Nov 4th, 2023'
---

# This is our first post

Here is a regular paragraph

here is a list:

- Item one
- Item two
- Item three

<CodeBlock></CodeBlock>

## This is an h2 element
```

Below are the following files

1. index.js for MDX
https://pastebin.com/5ZmCDtkf

2. articles/[slug]/page.js
https://pastebin.com/X0tGqsip

3. package.json
https://pastebin.com/iNCywNfK

4. mdx-components.js
https://pastebin.com/U8Bn739q

### Additional information

I am using NEXT JS 14.X

**Comment by jesperkc** [comment]
Looks like you never import the component in your mdx file. 

`import CodeBlock from './path to component'`

**Reply by jesperkc** [reply]
Ignore what I wrote. Have a look at https://github.com/hashicorp/next-mdx-remote#examples - it solved the issue for me

**Reply by Yusufzai** [reply]
Thank you for replying can it be the issue related between mdx 3 and next-mdx-remote

**Reply by Yusufzai** [reply]
Can you tell me what's your project nextjs version as currently I am using nextjs 14.X  with app directory

**Reply by jesperkc** [reply]
I'm using 14 as well with a combination of app and pages. I have all my mdx files in a "pages" folder in the root.

**Reply by namit343** [reply]
So installing this next-mdx-remote worked or what did you do exactly? I have the same issue, basically I was using a very old version of node, and build is failing now after many months on Vercel.

**Reply by Osmanity** [reply]
maybe it is time to update nodejs and other independences that follows

**Comment by nickfrosty** [comment]
I was getting a similar issue of custom MDX components no longer working after recently upgrading to NextJS 14 (from 13.x) while using `next-mdx-remote` and `@mdx-js/react`. Using only the `pages router`.

I still cannot figure out why exactly the issue was happening, but I was able to find a fix that works for me. Specifically by adjusting where the custom components props were passed within the `MDXProvider` and `MDXRemote` components.

My original code (that was working fine on NextJS v13.x and stopped working on v14.x):

```tsx
<MDXProvider components={customComponents}>
   <MDXRemote {...source} />
</MDXProvider>
```

The adjusted working code shifts the custom MDX components prop into the `MDXRemote` component:

```tsx
<MDXProvider>
   <MDXRemote {...source} components={customComponents} />
</MDXProvider>
```

**Reply by Osmanity** [reply]
was trying to find where i can make the changes you provided in the code but cant find the actual file where to changes the code as yours?

**Reply by Osmanity** [reply]
btw im using contentlayer is that something i can solve using your solution?

**Comment by shahid-shafi** [comment]
i am facing the same issue from last week, i have added custom components for example Note, h1 as below
none of them is working

```ts
import type { MDXComponents } from "mdx/types";

export function useMDXComponents(components: MDXComponents): MDXComponents {
  return {
    Note: ({ children }) => (
      <p className="text-2xl text-green-600">{children}</p>
    ),
    h1: ({ children }) => <h1 className="text-4xl text-red-600">{children}</h1>,
    ...components,
  };
}
```
and when i try to use in mdx file 
```mdx
---
title: "Test blog"
author: "John Doe"
description: "Lorem ipsum dolor sit amet consectetur adipisicing elit. Recusandae tenetur inventore nam libero eveniet. Modi consectetur, quia deserunt rerum doloribus, iste, explicabo similique praesentium ab sint quos incidunt ducimus. Ducimus?"
tags: ["ReactJS"]
created_at: "13-06-2024"
---
<Note>Just a note</Note>
```
this is the error i am getting:

Error: Expected component `Note` to be defined: you likely forgot to import, pass, or provide it.
It’s referenced in your code at `2:1-2:42`
